File: src/Stream|Sampling Problem/shuffle.cpp
/*
C++ swap is an in-place function 
array.size()
rand(): 0 ~ RAND_MAX
rand() % range + initial value
initialize random seed: srand(time(NULL))
*/
class Solution {
  public:
    vector<int> numbers;
    Solution(vector<int> nums) {
      numbers = nums; //this -> nums = nums;
    }
    
    vector<int> reset() {
      return numbers;
    }
  
    vector<int> shuffle() {
      vector<int> array = numbers; //vector<int> result(numbers);
      //srand(time(NULL)); 加上这句就过不了 ？？
      for(int i = 0; i < array.size(); i++) { 
        int candidate = i + rand()%(array.size() - i);
        swap(array[i], array[candidate]); 
      }
      return array;
    }
}
 

File: src/Stream|Sampling Problem/shuffle.java
import java.util.Random;

public class Solution {
  private int[] nums;
  private Random rand;
  
  public Solution(int[] nums) {
    this.nums = nums;
    rand = new Random();
  }
  
  public int[] reset() {
    return nums;
  }
  
  public int[] shuffle() {
    if(nums == null) {
      return null;
    }
    int[] array = nums.clone();
    for(int i = 0; i < array.length; i++) {
      int candidate = i + rand.nextInt(array.length - i);
      swap(array, candidate, i);
    }
  }
  
  private void swap(int[] array, int x, int y) {
    int tmp = array[x];
    array[x] = array[y];
    array[y] = tmp;
  }
}

